### PR TITLE
Fix order of parameters in AsContentElement and AsFrontendModule constructors

### LINKED ISSUE #4055: Problematic order of optional constructor parameters in AsContentElement/AsFrontendModule
### Affected version(s)

4.13

### Description

Our `AsContentElement` attribute currently looks like this:

https://github.com/contao/contao/blob/3691e1d7f9ee4eb35a9f5e05a3537dc633f94ffc/core-bundle/src/DependencyInjection/Attribute/AsContentElement.php#L23

Having the first non-optional parameter at the second place forces you to define at least two parameters (https://3v4l.org/Y7N4G#v8.1.2), but with the content element attribute, you typically only define the `category`, which then results in:

```
Contao\CoreBundle\DependencyInjection\Attribute\AsContentElement::__construct(): Argument #1 ($type) not passed
```

(Similar reasoning for `AsFrontendModule`.)

@aschempp What's the reason for this unusual order? Can we change this?

### PR DESCRIPTION
**Pull request by m-vo**
Fixes #4055

This PR swaps the first two parameters of the `AsContentElement` and `AsFrontendModule` constructors. This was not released yet, so no BC issue.

Background info: As of PHP8.0 having optional parameters before mandatory ones is deprecated. Here is an explanation why the current code did not trigger a deprecation warning and behaved just like two mandatory parameters: https://github.com/php/php-src/pull/5067#issuecomment-573872903

### CODE REVIEW

**Review by fritzmg — DISMISSED** [review]

**Review by ausi — DISMISSED** [review]

**Comment by aschempp** [comment]
I would kinda prefer @ausi's solution in https://github.com/contao/contao/issues/4055#issuecomment-1028927805
If I ever write an attribute with a type, I would never write 
```
#AsContentElement("texts", "foobar")
```

The goal should be to work the same as annotatinos did, so to use named paramters (I know we can't enforce that). Either write
```
#AsContentElement(category: 'texts')
```
or write
```
#AsContentElement("foobar", category: 'texts')
```
if you have a type. Making category default null and throw an exception would solve that.

**Comment by leofeyer** [comment]
@m-vo Will you update this PR accordingly?

**Comment by ausi** [comment]
I’d prefer something that works so that my IDE shows me if something is missing or what will happen.

IMO only the following two possibilities accomplish that:

```
__construct(string $category, string $type = null, …
__construct(string $type = null, string $category = 'texts',…
```

With `string $category = null` my IDE would suggest that this parameter is nullable and not required (which it is not at runtime then 😕)

**Comment by leofeyer** [comment]
I‘d be fine with `string $category = 'texts'` as well.

**Comment by m-vo** [comment]
But then you could omit all parameters and the element would appear in the text category (why text?). I'm ok with that, but I think I would prefer a solution without adding magic.

**Comment by leofeyer** [comment]
Why don‘t we just make `$type` a required parameter, too?

```php
__construct(string $type, string $category, …
```

The maker bundle will always set the type anyway:

https://github.com/contao/contao/blob/f577f1ebadebefd88eb44177ab4115a80fcad55a/maker-bundle/src/Resources/skeleton/content-element/ContentElement.tpl.php#L18-L20

**Comment by m-vo** [comment]
> Why don‘t we just make $type a required parameter, too?

In my opinion it's a good thing, that the type is inferred by default, because it promotes a common naming scheme. And I don't want to type it when creating fragments myself. :slightly_smiling_face:

**Comment by m-vo** [comment]
We should also discuss the usage of the variadic parameter.

**Comment by fritzmg** [comment]
I usually always end up using

```php
/**
 * @ContentElement(MyContentElement::TYPE, category="texts")
 */
class MyContentElement extends AbstractContentElementController
{
    public const TYPE = 'my_content_element';
}
```

so that I can reference the type in the DCA for example.

**Comment by m-vo** [comment]
> I usually always end up using
> 
> /**
>  * @ContentElement(MyContentElement::TYPE, category="texts")
>  */

:+1: for constants, that can be automatically refactored. Maybe the FQCN should get the default type in the future? 

Other aspect: We currently use the type to identify templates, but with Twig there is no point in injecting a `Template` in `getResponse()`. So either we build a special render method that, again, uses the type to identify a template or you would explicitly reference a template in `$this->render()` anyways and not need the type at all. This is also coupled to the discussion about template selection in the back end. 

If we want to alter the behavior in Contao 5, we could maybe already prepare this in 4.13 when using attributes.

**Review by leofeyer — CHANGES_REQUESTED** [review]
As discussed in the Contao call, we want to use `miscellaneous` as default category for both (!) content elements and front end modules. Also, we want to change `array ...$attributes` to `mixed ...$attributes`.

**Comment by m-vo** [comment]
The changes are now implemented as discussed.

Because we're now in a situation, where it's required to use named arguments to set arbitrary keys, I had to use some trickery in the tests, so that jobs < PHP8 would not result in a syntax error. If there is a better solution, let me know.

**Comment by ausi** [comment]
> If there is a better solution, let me know.

```php
new AsContentElement(...[
    'type' => 'content_element/text',
    'template' => 'a_template',
    'method' => 'aMethod',
    'renderer' => 'inline',
    'foo' => 'bar',
    'baz' => 42,
]);
```

https://3v4l.org/3mVUg

File: core-bundle/tests/DependencyInjection/ContaoCoreExtensionTest.php [hunk 1]
```diff
@@ -969,4 +969,11 @@ private function getContainerBuilder(array $params = null): ContainerBuilder
 
         return $container;
     }
+
+    private function skipTestIfAttributesAreNotSupported(): void
+    {
+        if (\PHP_VERSION_ID <= 80000) {
```
Comment (ausi) [inline core-bundle/tests/DependencyInjection/ContaoCoreExtensionTest.php]:
```suggestion
        if (\PHP_VERSION_ID < 80000) {
```

PHP 8.0.0 should be supported I guess?

Comment (m-vo) [reply core-bundle/tests/DependencyInjection/ContaoCoreExtensionTest.php]: Ah yes, my bad.

**Review by ausi — APPROVED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @m-vo.

**Pull request merged**